src/ivectorbin/ivector-cluster.cc

```diff
@@ -0,0 +1,141 @@
+// ivectorbin/ivector-cluster.cc
+
+// Copyright 2016  Matthew Maciejewski
+
+// See ../../COPYING for clarification regarding multiple authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+
+#include "base/kaldi-common.h"
+#include "util/common-utils.h"
+#include "util/stl-utils.h"
+#include "tree/cluster-utils.h"
+#include "tree/clusterable-classes.h"
+
+namespace kaldi {
+
+void PrepareSpeakerNums (const std::string utt2num_rxfilename,
+			 unordered_map<std::string, int32> *utt2num) {
+
+  Input ki(utt2num_rxfilename);
+
+  std::string line;
+  while (std::getline(ki.Stream(), line)) {
+    std::vector<std::string> split_line;
+    SplitStringToVector(line, " ", false, &split_line);
+    std::string utt = split_line[0],
+	num_str = split_line[1];
+
+    int32 num;
+    if (!ConvertStringToInteger(num_str, &num)) {
+      KALDI_ERR << "Invalid number of speakers in utt2num: " << line;
+    }
+
+    utt2num->insert(std::make_pair<std::string, int32>(utt, num));
+  }
+}
+
+}
+
+int main(int argc, char *argv[]) {
+  using namespace kaldi;
+  typedef kaldi::int32 int32;
+  typedef kaldi::int64 int64;
+  try {
+    const char *usage =
+	"Cluster iVectors for speaker diarization\n"
+	"Usage: ivector-cluster [options] <ivectors-rspecifier> "
+	"<ivector-ranges-rspecifier> <cluster-ranges-wspecifier>\n"
+	"e.g.: \n"
+	" ivector-cluster ark,t:ivectors.1.ark ark,t:ivector_ranges.1.ark \\\n"
+	"   ark,t:cluster_ranges.1.ark\n";
+
+    ParseOptions po(usage);
+    std::string ivector_weights_rspecifier,
+	        utt2num_rxfilename;
+    po.Register("ivector-weights-rspecifier", &ivector_weights_rspecifier,
+		"If supplied, uses the frames per ivector as weights "
+		"in the K-means clustering.");
```

Comment:
You might want to generalize this statement. The weights mustn't always be the number of frames per ivector.